src/NHibernate/Hql/Ast/ANTLR/Tree/SelectClause.cs

```diff
@@ -390,139 +437,286 @@ private void InitAliases(ISelectExpression[] selectExpressions)
 			}
 		}
 
-		private void RenderNonScalarSelects(ISelectExpression[] selectExpressions, FromClause currentFromClause)
+		private void RenderNonScalarSelects(
+			FromClause currentFromClause,
+			Dictionary<ISelectExpression, SelectClause> inheritedExpressions,
+			IList<FromElement> fetchedFromElements)
 		{
 			var appender = new ASTAppender(ASTFactory, this);
-			var nonscalarSize = selectExpressions.Count(e => !e.IsScalar);
+			var combinedFromElements = new List<FromElement>();
+			var processedElements = new HashSet<FromElement>();
+			RenderNonScalarIdentifiers(appender, processedElements, combinedFromElements, inheritedExpressions);
+			if (Walker.IsShallowQuery)
+			{
+				return;
+			}
+
+			// Append fetched elements
+			RenderFetchedNonScalarIdentifiers(appender, fetchedFromElements, processedElements, combinedFromElements);
+			if (currentFromClause.IsScalarSubQuery)
+			{
+				return;
+			}
+			
+			// Generate the property select tokens.
+			foreach (var fromElement in combinedFromElements)
+			{
+				RenderNonScalarProperties(appender, fromElement);
+			}
 
-			int j = 0;
-			foreach (var e in selectExpressions)
+			// Generate properties for fetched collections of components or values
+			var fromElements = currentFromClause.GetAllProjectionListTyped();
+			foreach (var fromElement in fromElements)
 			{
-				if (!e.IsScalar)
+				if (fromElement.IsCollectionOfValuesOrComponents &&
+					fromElement.IsFetch &&
+					processedElements.Add(fromElement))
 				{
-					FromElement fromElement = e.FromElement;
-					if (fromElement != null)
-					{
-						RenderNonScalarIdentifiers(fromElement, nonscalarSize, j, e, appender);
-						j++;
-					}
+					var suffix = Walker.GetSuffix(fromElement);
+					var fragment = fromElement.GetValueCollectionSelectFragment(suffix);
+					Append(appender, HqlSqlWalker.SQL_TOKEN, fragment);
 				}
 			}
+		}
 
-			if (!currentFromClause.IsSubQuery)
+		private void RenderNonScalarIdentifiers(
+			ASTAppender appender,
+			HashSet<FromElement> processedElements,
+			List<FromElement> combinedFromElements,
+			Dictionary<ISelectExpression, SelectClause> inheritedExpressions)
+		{
+			foreach (var e in NonScalarExpressions)
 			{
-				// Generate the property select tokens.
-				int k = 0;
-				foreach (var e in selectExpressions)
+				var fromElement = e.FromElement;
+				if (fromElement == null)
 				{
-					if (!e.IsScalar)
-					{
-						FromElement fromElement = e.FromElement;
-						if (fromElement != null)
-						{
-							RenderNonScalarProperties(appender, fromElement, nonscalarSize, k);
-							k++;
-						}
-					}
+					continue;
+				}
+
+				var node = (IASTNode) e;
+				if (processedElements.Add(fromElement))
```

Comment:
@maca88, this causes #3489 regression.
The query is
```
insert into Enrolment (Course, Student)
  select e.Course, e.Student from Enrolment e
```

At the second column, the from is already added, causing the execution to go into the else and not rendering the second column.

In debug, forcing the execution to jump directly to `RenderNonScalarIdentifiers` (so, not adding again the from to combinedFromElements) allows the test to succeed.

But surely the fix is not as simple as throwing away the else in all cases with something like:
```c#
if (processedElements.Add(fromElement))
{
	combinedFromElements.Add(fromElement);
}
RenderNonScalarIdentifiers(fromElement, inheritedExpressions.ContainsKey(e) ? null : e, appender);
```

Reply:
Doing that change seems to break only `HqlDuplicateEntitySelectionSubQuery` which looks like testing invalid cases, containing duplicate selection of entities. Why should we support theses cases?